portfolio/src/main/webapp/js/background.js

```diff
@@ -0,0 +1,375 @@
+      const prevPos = createVector(ball.pos.x, ball.pos.y);
+      ball.update(this);
+      this._updateHash(prevPos, ball);
+    });
+  }
+
+  /**
+   * Draws the balls in the tank.
+   */
+  draw() {
+    this._balls.forEach((ball, i) => ball.draw(this));
+  }
+
+  /**
+   * Adds a ball to the tank.
+   * @param {Ball} ball
+   */
+  addBall(ball) {
+    this._balls.push(ball);
+    this._placeInHash(this._getSpatialHashPos(ball.pos), ball);
+  }
+
+  /**
+   * Updates the spatial hashing position of a ball.
+   * @param {p5.Vector} prevPos
+   * @param {Ball} ball
+   * @private
+   */
+  _updateHash(prevPos, ball) {
+    const prevHashPos = this._getSpatialHashPos(prevPos);
+    const currentHashPos = this._getSpatialHashPos(ball.pos);
+    if (!prevHashPos.equals(currentHashPos)) {
+      if (!this._removeFromHash(prevHashPos, ball)) {
+        throw "Invalid previous ball position";
+      }
+      this._placeInHash(currentHashPos, ball);
+    }
+  }
+
+  /**
+   * Places a ball in spatial hash.
+   * @param {p5.Vector} hashPos
+   * @param {Ball} ball
+   * @private
+   */
+  _placeInHash(hashPos, ball) {
+    const hashKey = hashPos.x + "-" + hashPos.y;
+
+    if (!this._spatialHash[hashKey]) this._spatialHash[hashKey] = [];
+    this._spatialHash[hashKey].push(ball);
+  }
+
+  /**
+   * Removes a ball from the spatial hash.
+   * @param {p5.Vector} hashPos
+   * @param {Ball} ball
+   * @return {boolean}
+   * @private
+   */
+  _removeFromHash(hashPos, ball) {
+    const hashKey = hashPos.x + "-" + hashPos.y;
+
+    if (this._spatialHash[hashKey]) {
+      const ballIndex = this._spatialHash[hashKey].indexOf(ball);
+      if (ballIndex !== -1) {
+        this._spatialHash[hashKey].splice(ballIndex, 1);
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  /**
+   * Gets the spatial hashing position of a ball
+   * @param {p5.Vector} pos
+   * @return {p5.Vector}
+   * @private
+   */
+  _getSpatialHashPos(pos) {
+    return createVector(Math.floor(pos.x / this.staticDiameter),
+      Math.floor(pos.y / this.staticDiameter));
+  }
+
+  /**
+   * Get balls that could be affected by static effect. This is where spatial
+   * hashing is used to optimize.
+   * @param {Ball} ball
+   * @return {Array<Ball>}
+   */
+  getNearby(ball) {
+    // Approximate circular viewing limits to a square.
+    // This approximation holds because the spacial hash is divided into
+    // the diameter of the viewing circle, such that the circle can only
+    // ever be in four squares at once.
+
+    // Minimum x and y values are found at the left and top edges of the circle.
+    const xMin =
+      p5.Vector.sub(ball.pos, createVector(this.staticRadius, 0));
+    const yMin =
+      p5.Vector.sub(ball.pos, createVector(0, this.staticRadius));
+
+    const xMinHash = this._getSpatialHashPos(xMin).x;
+    const yMinHash = this._getSpatialHashPos(yMin).y;
+
+    // Can only ever occupy four squares, so iterate over
```

Comment:
Don't omit the subject from a sentence. You don't want a reader to have to guess what it is that can only occupy four squares. Change this to `Balls can only ever occupy four squares`